File: Data Structures & Algorithms/count-connected-components/submission-2.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        for e1, e2 in edges:
            graph[e1].append(e2)
            graph[e2].append(e1)
        visit = [False] * n

        def bfs(node):
            q = deque([node])
            visit[node] = True
            while q:
                cur = q.popleft()
                for nei in graph[cur]:
                    if not visit[nei]:
                        visit[nei] = True
                        q.append(nei)

        res = 0
        for node in range(n):
            if not visit[node]:
                bfs(node)
                res += 1
        return res

File: Data Structures & Algorithms/plus-one/submission-2.java
class Solution {
    public int[] plusOne(int[] digits) {
        digits[digits.length - 1] += 1;
        if(digits[digits.length - 1] < 10) return digits;

        digits[digits.length - 1] = 0;
        boolean carry = true;
        for(int i = digits.length - 2; i >= 0; --i) {
            if(carry) {
                if(digits[i] == 9) {
                    digits[i] = 0;
                }
                else {
                    digits[i]++;
                    carry = false;
                }
            }
        }
        if(carry) {
            int[] res = new int[digits.length + 1];
            res[0] = 1;
            return res;
        }
        return digits;
    }
}
